File: abstractFactory/python/OSXStore.py
from interfaces import AppInterface, GameInterface, StoreFactory

class OSXGame(GameInterface):
  __params = None
  PLATFORM = 'OSX'
  
  def __init__(self, name):
    self.__params = {
      'name': name,
      'type': 'game'
    }

    print('Game ', str(self.__params['name']), ' is created')    
    pass
  
  def play(self):
    print('Game ', str(self.__params['name']), ' is started!')
    pass
  
  def getInfo(self):
    return self.__params
  
  def printInformation(self):
    print('PLATFORM: ', str(OSXGame.PLATFORM))
    print(
      'GAME NAME: ', str(self.__params['name']), '\n'
      'TYPE: ', str(self.__params['type']), '\n'
    )
    pass

class OSXApp(AppInterface):
  __params = None
  PLATFORM = 'OSX'
  DEFAULTPRICE = 10

  def __init__(self, name):
    self.__params = {
      'name': name,
      'type': 'app',
      'price': OSXApp.DEFAULTPRICE
    }

    print('App ', str(self.__params['name']), ' is created')
    print(
      'APP NAME: ', str(self.__params['name']), '\n'
      'TYPE: ', str(self.__params['type']), '\n'
    )
    pass

  def getInfo(self):
    return self.__params
  
  def start(self):
    print('App ', str(self.__params['name']), ' is started')
    pass
  
  def buyApp(self, deposit):
    if deposit < self.__params['price']:
      raise ValueError(
        'your deposit less your: ', str(deposit),
        ' needed: ', str(self.__params['price'])
      )

    print(
      'App ', str(self.__params['name']),
      ' purchased for ', str(OSXApp.PLATFORM),
      ' written off ', str(self.__params['price'])
    )
    
    return True

  def printInformation(self):
    print('PLATFORM: ', str(OSXApp.PLATFORM))
    print(
      'APP NAME: ', str(self.__params['name']), '\n'
      'TYPE: ', str(self.__params['type']), '\n'
      'APP PRICE: ', str(self.__params['price']), '\n'
    )
    pass

class OSXStore(StoreFactory):
  def createGame(self, name):
    return OSXGame(name)
  def createApplication(self, name):
    return OSXApp(name)

File: abstractFactory/cpp/osx_store.cpp
#ifndef osx_store
#define osx_store

#include <tuple>
#include <string>
#include <iostream>
#include <stdexcept>
#include "interfaces.hpp"

using namespace std;

class OSXGame : public GameInterface
{
  private:
    std::string name;
    std::string type = "game";
  public:
    OSXGame(std::string name) {
      this->name = name;
    }

    std::tuple<std::string, std::string> getInfo() {
      return std::make_tuple(this->name, this->type);
    }

    void play() {
      cout << "Game " << this->name << " runing" << endl;
    }

    void printInformation() {
      cout << "PLATFORM: Windows" << endl;
      cout << "NAME:     " << this->name << endl;
      cout << "TYPE:     " << this->type << endl;
    }
};

class OSXApp : public AppInterface
{
  private:
    std::string name;
    std::string type = "app";
    int default_price = 10;
  public:
    OSXApp(std::string name) {
      this->name = name;
    }

    std::tuple<std::string, std::string, int> getInfo() {
      return std::make_tuple(this->name, this->type, this->default_price);
    }

    bool buyApp(int deposit) {
      if (this->default_price != deposit) {
        throw std::invalid_argument("deposit less");
      }

      cout << "App " << this->name << " buyed success" << endl;
      return true;
    }

    void start() {
      cout << "App " << this->name << " started" << endl;
    }

    void printInformation() {
      cout << "PLATFORM: Windows" << endl;
      cout << "NAME:     " << this->name << endl;
      cout << "TYPE:     " << this->type << endl;
      cout << "PRICE:    " << this->default_price << endl;
    }
};

class OSXStore : public StoreFactory
{
  public:
    AppInterface * createApp(std::string name) {
      return new OSXApp(name);
    }

    GameInterface * createGame(std::string name) {
      return new OSXGame(name);
    }
};

#endif